Remove usage of `assume_checked()`

Follow-up to #2356

Bitcoin v30 introduced the concept of checked and unchecked addresses. While migrating over from v0.29 to v0.30, I called `assume_checked()` wherever necessary to make the compiler happy. Now that the migration is complete, we should remove these calls and instead use [`require_network()`](https://docs.rs/bitcoin/0.30.2/bitcoin/address/struct.Address.html#method.require_network).

Edit: See [this comment](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/bitcoin/CHANGELOG.md#changes-to-and-involving-the-network-type) in the changelog for `bitcoin` v0.32.0 which indicates that the work here might not be that important.

am getting the following errors when i ran 

just check command

```

error[E0658]: use of unstable library feature 'lazy_cell'
  --> fedimint-bitcoind/src/lib.rs:14:22
   |
14 | use std::sync::{Arc, LazyLock, Mutex};
   |                      ^^^^^^^^
   |
   = note: see issue #109736 <https://github.com/rust-lang/rust/issues/109736> for more information

error[E0658]: use of unstable library feature 'lazy_cell'
  --> fedimint-bitcoind/src/lib.rs:50:30
   |
50 | static BITCOIN_RPC_REGISTRY: LazyLock<Mutex<BTreeMap<String, DynBitcoindRpcFactory>>> =
   |                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: see issue #109736 <https://github.com/rust-lang/rust/issues/109736> for more information

error[E0658]: use of unstable library feature 'lazy_cell'
  --> fedimint-bitcoind/src/lib.rs:51:5
   |
51 |     LazyLock::new(|| {
   |     ^^^^^^^^
   |
   = note: see issue #109736 <https://github.com/rust-lang/rust/issues/109736> for more information

error[E0658]: use of unstable library feature 'lazy_cell'
  --> fedimint-bitcoind/src/lib.rs:51:5
   |
51 |     LazyLock::new(|| {
   |     ^^^^^^^^^^^^^
   |
   = note: see issue #109736 <https://github.com/rust-lang/rust/issues/109736> for more information

```

@tvolk131

--- a/devimint/src/external.rs
+++ b/devimint/src/external.rs
@@ -131,7 +131,9 @@ impl Bitcoind {
         if !skip_setup {
             // mine blocks
             let blocks = 101;
-            let address = block_in_place(|| client.get_new_address(None, None))?.assume_checked();
+            let address = block_in_place(|| client.get_new_address(None, None))?
+                .require_network(bitcoin::Network::Regtest)
+                .expect("Devimint always runs in regtest");
             debug!(target: LOG_DEVIMINT, blocks_num=blocks, %address, "Mining blocks to address");
             block_in_place(|| {
                 client
@@ -225,7 +227,12 @@ impl Bitcoind {
         let tx = self
             .wallet_client()
             .await?
-            .send_to_address(bitcoin::Address::from_str(&addr)?.assume_checked(), amount)
+            .send_to_address(
+                bitcoin::Address::from_str(&addr)?
+                    .require_network(bitcoin::Network::Regtest)
+                    .expect("Devimint always runs in regtest"),
+                amount,
+            )
             .await?;
         Ok(tx)
     }
@@ -327,7 +334,8 @@ impl Bitcoind {
         let client = self.wallet_client().await?.clone();
         let addr = spawn_blocking(move || client.client.get_new_address(None, None))
             .await??
-            .assume_checked();
+            .require_network(bitcoin::Network::Regtest)
+            .expect("Devimint always runs in regtest");
         Ok(addr)
     }
 

--- a/fedimint-cli/src/client.rs
+++ b/fedimint-cli/src/client.rs
@@ -554,15 +554,14 @@ pub async fn handle_command(
         }
         ClientCmd::Withdraw { amount, address } => {
             let wallet_module = client.get_first_module::<WalletClientModule>()?;
+            let address = address.require_network(wallet_module.get_network())?;
             let (amount, fees) = match amount {
                 // If the amount is "all", then we need to subtract the fees from
                 // the amount we are withdrawing
                 BitcoinAmountOrAll::All => {
                     let balance =
                         bitcoin::Amount::from_sat(client.get_balance().await.msats / 1000);
-                    let fees = wallet_module
-                        .get_withdraw_fees(address.clone(), balance)
-                        .await?;
+                    let fees = wallet_module.get_withdraw_fees(&address, balance).await?;
                     let amount = balance.checked_sub(fees.amount());
                     if amount.is_none() {
                         bail!("Not enough funds to pay fees");
@@ -571,9 +570,7 @@ pub async fn handle_command(
                 }
                 BitcoinAmountOrAll::Amount(amount) => (
                     amount,
-                    wallet_module
-                        .get_withdraw_fees(address.clone(), amount)
-                        .await?,
+                    wallet_module.get_withdraw_fees(&address, amount).await?,
                 ),
             };
             let absolute_fees = fees.amount();
@@ -583,7 +580,7 @@ pub async fn handle_command(
                 "Attempting withdraw with fees: {fees:?}"
             );
 
-            let operation_id = wallet_module.withdraw(address, amount, fees, ()).await?;
+            let operation_id = wallet_module.withdraw(&address, amount, fees, ()).await?;
 
             let mut updates = wallet_module
                 .subscribe_withdraw_updates(operation_id)

--- a/fedimint-core/src/encoding/btc.rs
+++ b/fedimint-core/src/encoding/btc.rs
@@ -203,10 +203,13 @@ impl Decodable for bitcoin::Amount {
 impl Encodable for bitcoin::Address<NetworkUnchecked> {
     fn consensus_encode<W: std::io::Write>(&self, writer: &mut W) -> Result<usize, Error> {
         let mut len = 0;
-        len += NetworkLegacyEncodingWrapper(get_network_for_address(self.as_unchecked()))
-            .consensus_encode(writer)?;
+        len +=
+            NetworkLegacyEncodingWrapper(get_network_for_address(self)).consensus_encode(writer)?;
         len += self
             .clone()
+            // We need an `Address<NetworkChecked>` in order to get the script pubkey.
+            // Calling `assume_checked` is generally a bad idea, but it's safe here where we're
+            // encoding the address because addresses are always decoded as unchecked.
             .assume_checked()
             .script_pubkey()
             .consensus_encode(writer)?;
@@ -225,7 +228,7 @@ impl Decodable for bitcoin::Address<NetworkUnchecked> {
         let address = bitcoin::Address::from_script(&script_pk, network)
             .map_err(|e| DecodeError::new_custom(e.into()))?;
 
-        Ok(address.as_unchecked().clone())
+        Ok(address.into_unchecked())
     }
 }
 

--- a/gateway/ln-gateway/src/lib.rs
+++ b/gateway/ln-gateway/src/lib.rs
@@ -63,7 +63,9 @@ use fedimint_core::secp256k1::PublicKey;
 use fedimint_core::task::{sleep, TaskGroup, TaskHandle, TaskShutdownToken};
 use fedimint_core::time::duration_since_epoch;
 use fedimint_core::util::{SafeUrl, Spanned};
-use fedimint_core::{fedimint_build_code_version_env, Amount, BitcoinAmountOrAll};
+use fedimint_core::{
+    fedimint_build_code_version_env, get_network_for_address, Amount, BitcoinAmountOrAll,
+};
 use fedimint_eventlog::{DBTransactionEventLogExt, EventLogId};
 use fedimint_ln_common::config::{GatewayFee, LightningClientConfig};
 use fedimint_ln_common::contracts::Preimage;
@@ -886,6 +888,15 @@ impl Gateway {
             address,
             federation_id,
         } = payload;
+
+        let address_network = get_network_for_address(&address);
+        let gateway_network = self.gateway_config.read().await.network.0;
+        let Ok(address) = address.require_network(gateway_network) else {
+            return Err(AdminGatewayError::WithdrawError {
+                failure_reason: format!("Gateway is running on network {gateway_network}, but provided withdraw address is for network {address_network}")
+            });
+        };
+
         let client = self.select_client(federation_id).await?;
         let wallet_module = client.value().get_first_module::<WalletClientModule>()?;
 
@@ -896,9 +907,7 @@ impl Gateway {
             BitcoinAmountOrAll::All => {
                 let balance =
                     bitcoin::Amount::from_sat(client.value().get_balance().await.msats / 1000);
-                let fees = wallet_module
-                    .get_withdraw_fees(address.clone(), balance)
-                    .await?;
+                let fees = wallet_module.get_withdraw_fees(&address, balance).await?;
                 let withdraw_amount = balance.checked_sub(fees.amount());
                 if withdraw_amount.is_none() {
                     return Err(AdminGatewayError::WithdrawError {
@@ -911,15 +920,11 @@ impl Gateway {
             }
             BitcoinAmountOrAll::Amount(amount) => (
                 amount,
-                wallet_module
-                    .get_withdraw_fees(address.clone(), amount)
-                    .await?,
+                wallet_module.get_withdraw_fees(&address, amount).await?,
             ),
         };
 
-        let operation_id = wallet_module
-            .withdraw(address.clone(), amount, fees, ())
-            .await?;
+        let operation_id = wallet_module.withdraw(&address, amount, fees, ()).await?;
         let mut updates = wallet_module
             .subscribe_withdraw_updates(operation_id)
             .await?
@@ -928,10 +933,7 @@ impl Gateway {
         while let Some(update) = updates.next().await {
             match update {
                 WithdrawState::Succeeded(txid) => {
-                    info!(
-                        "Sent {amount} funds to address {}",
-                        address.assume_checked()
-                    );
+                    info!("Sent {amount} funds to address {}", address);
                     return Ok(WithdrawResponse { txid, fees });
                 }
                 WithdrawState::Failed(e) => {
@@ -1369,13 +1371,20 @@ impl Gateway {
     pub async fn handle_get_ln_onchain_address_msg(&self) -> AdminResult<Address> {
         let context = self.get_lightning_context().await?;
         let response = context.lnrpc.get_ln_onchain_address().await?;
-        Address::from_str(&response.address)
-            .map(Address::assume_checked)
-            .map_err(|e| {
-                AdminGatewayError::Lightning(LightningRpcError::InvalidMetadata {
-                    failure_reason: e.to_string(),
-                })
+
+        let address = Address::from_str(&response.address).map_err(|e| {
+            AdminGatewayError::Lightning(LightningRpcError::InvalidMetadata {
+                failure_reason: e.to_string(),
             })
+        })?;
+
+        let network = self.gateway_config.read().await.network.0;
+
+        address.require_network(network).map_err(|e| {
+            AdminGatewayError::Lightning(LightningRpcError::InvalidMetadata {
+                failure_reason: e.to_string(),
+            })
+        })
     }
 
     /// Instructs the Gateway's Lightning node to open a channel to a peer

--- a/modules/fedimint-wallet-client/src/lib.rs
+++ b/modules/fedimint-wallet-client/src/lib.rs
@@ -25,7 +25,7 @@ use std::future;
 use std::sync::Arc;
 use std::time::SystemTime;
 
-use anyhow::{anyhow, bail, ensure, Context as AnyhowContext};
+use anyhow::{anyhow, bail, Context as AnyhowContext};
 use async_stream::stream;
 use backup::WalletModuleBackup;
 use bitcoin::address::NetworkUnchecked;
@@ -549,13 +549,11 @@ impl WalletClientModule {
     /// The caller should be prepared to retry with a new fee estimate.
     pub async fn get_withdraw_fees(
         &self,
-        address: bitcoin::Address<NetworkUnchecked>,
+        address: &bitcoin::Address,
         amount: bitcoin::Amount,
     ) -> anyhow::Result<PegOutFees> {
-        check_address(&address, self.cfg().network.0)?;
-
         self.module_api
-            .fetch_peg_out_fees(&address.assume_checked(), amount)
+            .fetch_peg_out_fees(address, amount)
             .await?
             .context("Federation didn't return peg-out fees")
     }
@@ -568,13 +566,11 @@ impl WalletClientModule {
     pub fn create_withdraw_output(
         &self,
         operation_id: OperationId,
-        address: &bitcoin::Address<NetworkUnchecked>,
+        address: bitcoin::Address,
         amount: bitcoin::Amount,
         fees: PegOutFees,
     ) -> anyhow::Result<ClientOutputBundle<WalletOutput, WalletClientStates>> {
-        check_address(address, self.cfg().network.0)?;
-
-        let output = WalletOutput::new_v0_peg_out(address.clone(), amount, fees);
+        let output = WalletOutput::new_v0_peg_out(address, amount, fees);
 
         let amount = output.maybe_v0_ref().expect("v0 output").amount().into();
 
@@ -677,8 +673,7 @@ impl WalletClientModule {
                             WalletCommonInit::KIND.as_str(),
                             WalletOperationMeta {
                                 variant: WalletOperationMetaVariant::Deposit {
-                                    // TODO(bitcoin 0.32): use as_unchecked
-                                    address: address.to_string().parse().expect("can be parsed"),
+                                    address: address.clone().into_unchecked(),
                                     tweak_idx: Some(tweak_idx),
                                     expires_at: None,
                                 },
@@ -744,6 +739,8 @@ impl WalletClientModule {
             bail!("Operation is not a deposit operation");
         };
 
+        let address = address.require_network(self.cfg().network.0)?;
+
         // The old deposit operations don't have tweak_idx set
         let Some(tweak_idx) = tweak_idx else {
             // In case we are dealing with an old deposit that still uses state machines we
@@ -771,7 +768,7 @@ impl WalletClientModule {
         Ok(self.client_ctx.outcome_or_updates(&operation, operation_id, || {
             let stream_rpc = self.rpc.clone();
             let stream_client_ctx = self.client_ctx.clone();
-            let stream_script_pub_key = address.assume_checked().script_pubkey();
+            let stream_script_pub_key = address.script_pubkey();
 
             stream! {
                 yield DepositStateV2::WaitingForTransaction;
@@ -1009,7 +1006,7 @@ impl WalletClientModule {
     /// acknowledged by the user since it can be unexpectedly high.
     pub async fn withdraw<M: Serialize + MaybeSend + MaybeSync>(
         &self,
-        address: bitcoin::Address<NetworkUnchecked>,
+        address: &bitcoin::Address,
         amount: bitcoin::Amount,
         fee: PegOutFees,
         extra_meta: M,
@@ -1018,7 +1015,7 @@ impl WalletClientModule {
             let operation_id = OperationId(thread_rng().gen());
 
             let withdraw_output =
-                self.create_withdraw_output(operation_id, &address, amount, fee)?;
+                self.create_withdraw_output(operation_id, address.clone(), amount, fee)?;
             let tx_builder = TransactionBuilder::new()
                 .with_outputs(self.client_ctx.make_client_outputs(withdraw_output));
 
@@ -1030,7 +1027,7 @@ impl WalletClientModule {
                     WalletCommonInit::KIND.as_str(),
                     |_, change| WalletOperationMeta {
                         variant: WalletOperationMetaVariant::Withdraw {
-                            address: address.clone(),
+                            address: address.clone().into_unchecked(),
                             amount,
                             fee,
                             change,
@@ -1146,15 +1143,6 @@ impl WalletClientModule {
                 }
             }))
     }
-}
-
-fn check_address(address: &Address<NetworkUnchecked>, network: Network) -> anyhow::Result<()> {
-    ensure!(
-        address.is_valid_for_network(network),
-        "Address isn't compatible with the federation's network: {network:?}"
-    );
-
-    Ok(())
 }
 
 /// Returns the child index to derive the next peg-in tweak key from.

--- a/modules/fedimint-wallet-common/src/lib.rs
+++ b/modules/fedimint-wallet-common/src/lib.rs
@@ -386,12 +386,12 @@ extensible_associated_module_type!(
 
 impl WalletOutput {
     pub fn new_v0_peg_out(
-        recipient: Address<NetworkUnchecked>,
+        recipient: Address,
         amount: bitcoin::Amount,
         fees: PegOutFees,
     ) -> WalletOutput {
         WalletOutput::V0(WalletOutputV0::PegOut(PegOut {
-            recipient,
+            recipient: recipient.into_unchecked(),
             amount,
             fees,
         }))

--- a/modules/fedimint-wallet-server/src/lib.rs
+++ b/modules/fedimint-wallet-server/src/lib.rs
@@ -868,6 +868,9 @@ impl ServerModule for Wallet {
 
                     let tx = module.offline_wallet().create_tx(
                         bitcoin::Amount::from_sat(sats),
+                        // Note: While calling `assume_checked()` is generally unwise, it's fine
+                        // here since we're only returning a fee estimate, and we would still
+                        // reject a transaction with the wrong network upon attempted peg-out.
                         address.assume_checked().script_pubkey(),
                         vec![],
                         module.available_utxos(&mut context.dbtx().into_nc()).await,
@@ -1410,6 +1413,9 @@ impl Wallet {
         match output {
             WalletOutputV0::PegOut(peg_out) => self.offline_wallet().create_tx(
                 peg_out.amount,
+                // Note: While calling `assume_checked()` is generally unwise, checking the
+                // network here could be a consensus-breaking change. Ignoring the network
+                // is fine here since we validate it in `process_output()`.
                 peg_out.recipient.clone().assume_checked().script_pubkey(),
                 vec![],
                 self.available_utxos(dbtx).await,

--- a/modules/fedimint-wallet-tests/tests/tests.rs
+++ b/modules/fedimint-wallet-tests/tests/tests.rs
@@ -282,16 +282,12 @@ async fn on_chain_peg_in_and_peg_out_happy_case() -> anyhow::Result<()> {
     // Peg-out test, requires block to recognize change UTXOs
     let address = bitcoin.get_new_address().await;
     let peg_out = bsats(PEG_OUT_AMOUNT_SATS);
-    let fees = wallet_module
-        .get_withdraw_fees(address.as_unchecked().clone(), peg_out)
-        .await?;
+    let fees = wallet_module.get_withdraw_fees(&address, peg_out).await?;
     assert_eq!(
         fees.total_weight, 871,
         "stateless wallet should have constructed a tx with a total weight=871"
     );
-    let op = wallet_module
-        .withdraw(address.as_unchecked().clone(), peg_out, fees, ())
-        .await?;
+    let op = wallet_module.withdraw(&address, peg_out, fees, ()).await?;
 
     let balance_after_peg_out =
         sats(PEG_IN_AMOUNT_SATS - PEG_OUT_AMOUNT_SATS - fees.amount().to_sat());
@@ -429,9 +425,7 @@ async fn peg_out_fail_refund() -> anyhow::Result<()> {
     };
 
     let wallet_module = client.get_first_module::<WalletClientModule>()?;
-    let op = wallet_module
-        .withdraw(address.as_unchecked().clone(), peg_out, fees, ())
-        .await?;
+    let op = wallet_module.withdraw(&address, peg_out, fees, ()).await?;
     assert_eq!(
         balance_sub.next().await.unwrap(),
         sats(PEG_IN_AMOUNT_SATS - PEG_OUT_AMOUNT_SATS)
@@ -469,12 +463,8 @@ async fn rbf_withdrawals_are_rejected() -> anyhow::Result<()> {
     let address = bitcoin.get_new_address().await;
     let peg_out = bsats(PEG_OUT_AMOUNT_SATS);
     let wallet_module = client.get_first_module::<WalletClientModule>()?;
-    let fees = wallet_module
-        .get_withdraw_fees(address.as_unchecked().clone(), peg_out)
-        .await?;
-    let op = wallet_module
-        .withdraw(address.as_unchecked().clone(), peg_out, fees, ())
-        .await?;
+    let fees = wallet_module.get_withdraw_fees(&address, peg_out).await?;
+    let op = wallet_module.withdraw(&address, peg_out, fees, ()).await?;
 
     let sub = wallet_module.subscribe_withdraw_updates(op).await?;
     let mut sub = sub.into_stream();
@@ -568,10 +558,10 @@ async fn peg_outs_must_wait_for_available_utxos() -> anyhow::Result<()> {
     let peg_out1 = PEG_OUT_AMOUNT_SATS;
     let wallet_module = client.get_first_module::<WalletClientModule>()?;
     let fees1 = wallet_module
-        .get_withdraw_fees(address.as_unchecked().clone(), bsats(peg_out1))
+        .get_withdraw_fees(&address, bsats(peg_out1))
         .await?;
     let op = wallet_module
-        .withdraw(address.as_unchecked().clone(), bsats(peg_out1), fees1, ())
+        .withdraw(&address, bsats(peg_out1), fees1, ())
         .await?;
     let balance_after_peg_out =
         sats(PEG_IN_AMOUNT_SATS - PEG_OUT_AMOUNT_SATS - fees1.amount().to_sat());
@@ -594,7 +584,7 @@ async fn peg_outs_must_wait_for_available_utxos() -> anyhow::Result<()> {
     let address = bitcoin.get_new_address().await;
     let peg_out2 = PEG_OUT_AMOUNT_SATS;
     let fees2 = wallet_module
-        .get_withdraw_fees(address.as_unchecked().clone(), bsats(peg_out2))
+        .get_withdraw_fees(&address, bsats(peg_out2))
         .await;
     // Must fail because change UTXOs are still being confirmed
     assert!(fees2.is_err());
@@ -604,10 +594,10 @@ async fn peg_outs_must_wait_for_available_utxos() -> anyhow::Result<()> {
     await_consensus_to_catch_up(&client, current_block + 1).await?;
     // Now change UTXOs are available and we can peg-out again
     let fees2 = wallet_module
-        .get_withdraw_fees(address.as_unchecked().clone(), bsats(peg_out2))
+        .get_withdraw_fees(&address, bsats(peg_out2))
         .await?;
     let op = wallet_module
-        .withdraw(address.as_unchecked().clone(), bsats(peg_out2), fees2, ())
+        .withdraw(&address, bsats(peg_out2), fees2, ())
         .await?;
     let sub = wallet_module.subscribe_withdraw_updates(op).await?;
     let mut sub = sub.into_stream();
@@ -825,12 +815,8 @@ async fn construct_wallet_summary() -> anyhow::Result<()> {
     //   - after the tx is mined and finalized
     let address = bitcoin.get_new_address().await;
     let peg_out = bsats(PEG_OUT_AMOUNT_SATS);
-    let fees = wallet_module
-        .get_withdraw_fees(address.as_unchecked().clone(), peg_out)
-        .await?;
-    let op = wallet_module
-        .withdraw(address.as_unchecked().clone(), peg_out, fees, ())
-        .await?;
+    let fees = wallet_module.get_withdraw_fees(&address, peg_out).await?;
+    let op = wallet_module.withdraw(&address, peg_out, fees, ()).await?;
 
     let sub = wallet_module.subscribe_withdraw_updates(op).await?;
     let mut sub = sub.into_stream();

